File: day3/part1/solution.c
/* Advent of Code Day 3 - part1 */
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../../utils/utils.h"

int main() {
    printf("Advent of Code Day 3 - part1\n");
    char *file_input = load_input("../input.txt");
    int total = 0;

    char *ptr = file_input;
    while(*ptr != '\0')
    {
        if (strncmp(ptr, "mul(", 4) != 0)
        {
            ptr++;
            continue;
        }
        ptr += 4;
        char *start = ptr;
        int length = 0;
        while(isdigit(*ptr))
        {
            length++;
            ptr++;
        }
        if(*ptr != ',')
        {
            continue;
        }
        ptr++;
        char substring1[length+1];
        strncpy(substring1, start, length);
        substring1[length] = '\0';
        int num1 = atoi(substring1);

        start = ptr;
        length = 0;
        while(isdigit(*ptr))
        {
            length++;
            ptr++;
        }
        if(*ptr != ')')
        {
            continue;
        }
        ptr++;
        char substring2[length+1];
        strncpy(substring2, start, length);
        substring2[length] = '\0';
        int num2 = atoi(substring2);
        total += num1 * num2;
    }

    printf("%i\n", total);
    free(file_input);
    return 0;
}

File: day3/part1/solution.py
# Advent of Code Day 3 - part1

import re


def main():
    print('Advent of Code Day 3 - part1')

    with open('../input.txt', 'r') as file:
        data = file.read()

    regex = r'mul\((\d+),(\d+)\)'
    datamatch = re.findall(regex, data)
    total = 0
    extract = [(int(num1), int(num2)) for num1, num2 in datamatch]
    for i in range(len(extract)):
        total += extract[i][0] * extract[i][1]
    print(total)


if __name__ == '__main__':
    main()
